container/thread-safe/map.go

```diff
@@ -0,0 +1,58 @@
+// Package threadsafe contains generic containers that are
+// protected either by Mutexes or atomics underneath the hood.
+package threadsafe
+
+import "sync"
+
+// Map implements a simple thread-safe map protected by a mutex.
+type Map[K comparable, V any] struct {
+	items map[K]V
+	lock  sync.RWMutex
+}
+
+// NewThreadSafeMap returns a thread-safe map instance from a normal map.
+func NewThreadSafeMap[K comparable, V any](m map[K]V) *Map[K, V] {
+	return &Map[K, V]{
+		items: m,
+	}
+}
+
+// Keys returns the keys of a thread-safe map.
+func (m *Map[K, V]) Keys() []K {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	r := make([]K, 0, len(m.items))
+	for k := range m.items {
+		r = append(r, k)
```

Comment:
Depending on what type `K` is you might run into the for loop reference bug, where the reference changes with each iteration and in the slice. Since this is generic and to safeguard against all types, you should copy `k` before appending it to `r`

Reply:
let's just get rid of this method because there is no way of copying k generically unless it implements some Copy method

Reply:
sorry not a deep copy, but just copy the value . Ex:
```
copiedK := k
r = append(r, copiedK)
```

Reply:
Pushed